src/Workspaces/Core/Portable/CodeStyle/CodeStyleOption.cs

```diff
@@ -3,194 +3,56 @@
-    /// Represents a code style option and an associated notification option.  Supports
-    /// being instantiated with T as a <see cref="bool"/> or an <c>enum type</c>.
-    /// 
-    /// CodeStyleOption also has some basic support for migration a <see cref="bool"/> option
-    /// forward to an <c>enum type</c> option.  Specifically, if a previously serialized
-    /// bool-CodeStyleOption is then deserialized into an enum-CodeStyleOption then 'false' 
-    /// values will be migrated to have the 0-value of the enum, and 'true' values will be
-    /// migrated to have the 1-value of the enum.
-    /// 
-    /// Similarly, enum-type code options will serialize out in a way that is compatible with 
-    /// hosts that expect the value to be a boolean.  Specifically, if the enum value is 0 or 1
-    /// then those values will write back as false/true.
-    /// </summary>
+    /// <inheritdoc cref="CodeStyleOption2{T}"/>
     public class CodeStyleOption<T> : ICodeStyleOption, IEquatable<CodeStyleOption<T>>
     {
+        private readonly CodeStyleOption2<T> _codeStyleOptionImpl;
         public static CodeStyleOption<T> Default => new CodeStyleOption<T>(default, NotificationOption.Silent);
 
-        private const int SerializationVersion = 1;
-
-        private NotificationOption _notification;
+        internal CodeStyleOption(CodeStyleOption2<T> codeStyleOptionImpl)
+        {
+            _codeStyleOptionImpl = codeStyleOptionImpl;
+        }
 
         public CodeStyleOption(T value, NotificationOption notification)
+            : this(new CodeStyleOption2<T>(value, notification ?? throw new ArgumentNullException(nameof(notification))))
         {
-            Value = value;
-            _notification = notification ?? throw new ArgumentNullException(nameof(notification));
         }
 
-        public T Value { get; set; }
+        public T Value
+        {
+            get => _codeStyleOptionImpl.Value;
+            set => _codeStyleOptionImpl.Value = value;
+        }
 
         object ICodeStyleOption.Value => this.Value;
+        NotificationOption2 ICodeStyleOption.Notification => _codeStyleOptionImpl.Notification;
         ICodeStyleOption ICodeStyleOption.WithValue(object value) => new CodeStyleOption<T>((T)value, Notification);
-        ICodeStyleOption ICodeStyleOption.WithNotification(NotificationOption notification) => new CodeStyleOption<T>(Value, notification);
-
-        private int EnumValueAsInt32 => (int)(object)Value;
+        ICodeStyleOption ICodeStyleOption.WithNotification(NotificationOption2 notification) => new CodeStyleOption<T>(Value, notification);
 
         public NotificationOption Notification
         {
-            get => _notification;
-            set => _notification = value ?? throw new ArgumentNullException(nameof(value));
+            get => _codeStyleOptionImpl.Notification;
+            set => _codeStyleOptionImpl.Notification = value ?? throw new ArgumentNullException(nameof(value));
```

Comment:
```suggestion
            set => _codeStyleOptionImpl.Notification = value;
```

Reply:
Retaining the existing public API semantics for this PR.

Reply:
The pass-through setter has these same semantics right?